texts/0005-api-refactor.md

```diff
@@ -0,0 +1,79 @@
+
+## Platform Considerations
+
+We can improve the existing desktop APIs, but it is not required to do so. The main goal of this refactor is to actually implement existing functionality on mobile, though improvements are always available and are being discussed for instance on the [capability-based API issue](https://github.com/tauri-apps/tauri/issues/6107).
+
+## Security Considerations
+
+With each API as a separate package, it will be easier to keep the security guarantees from previous audits. Changes will be easier to track and audit.
+
+## Performance Considerations
+
+Performance is not impacted by this change. Existing desktop code runs the same way, and mobile APIs are executed in the native layer so it can leverage the most performant way of doing things.
+
+## Community Considerations
+
+This refactor introduces a massive breaking change. All existing code using the `tauri::api` module or the `@tauri-apps/api` package would break. To mitigate this and simplify the migration path to v2, I propose we move all existing functionality (both the `tauri::api` module and the current Tauri endpoints) to a `legacy` plugin that can be enabled with a Cargo feature. With this plugin, an existing Tauri v1 application could start using the v2 functionalities by just enabling the Cargo feature and changing the `tauri::api` imports to `tauri::legacy`. This solution can also be used for other breaking changes for v2 and beyond, as existing functions that have incompatible new signatures could be moved to a Rust trait that can be imported by the developers to migrate major versions without touching their existing codebase.
+
+# Drawbacks
+
+This change introduces a lot of new packages. Each plugin must have a crate and an NPM package exposing the webview API usage. Keeping both versions in sync is necessary for compatibility, so it introduces complexity for end users. To help users, we are proposing a [`tauri add` CLI command](https://github.com/tauri-apps/tauri/issues/6505).
+
+# Rationale and alternatives
+
+- This design is good for maintainability as it separates application APIs from the core crate. Additionally, it simplifies the mess associated with target detection for the huge list of platforms we must support: macOS, Linux, Windows, Android and iOS.
+- It reduces the current feature flags hell with the allowlist Cargo features. Currently, the tauri crate has over 70 Cargo features related to the allowlist. These flags would be moved to each API plugin crate.
+
+# Prior art
+
+For possible mobile-specific plugins, we can learn a lot from the official Capacitor [plugin collection](https://github.com/ionic-team/capacitor-plugins).
+
+# Unresolved questions
+
+How should we handle versioning? Should we keep minors in sync with the core crate so it is easier to determine incompatibilities? Syncing can be problematic when we have 20+ plugins and we only want to minor bump one of them...
```

Comment:
This may be better to circle back around to after we have the release discussion in the next working group call?